### ISSUE TITLE
Git reset fails when file with the name 'origin' is present in the repository

### ISSUE THREAD

**Issue by silwol** [ansible/ansible#22455]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
git

##### ANSIBLE VERSION

```
ansible 2.2.1.0
  config file = /home/user/Projects/ansible_site/ansible.cfg
  configured module search path = Default w/o overrides
```

##### OS / ENVIRONMENT

Debian Stretch, with Ansible installed in a python venv.

##### SUMMARY
```
/usr/bin/git reset --hard origin
```
can be an ambiguous call when a file named "origin" is present in the repository.

Appending "--" to the command would make it robust.
See http://stackoverflow.com/questions/26349191/fatal-ambiguous-argument-branch-name-both-revision-and-filename for further details.

##### STEPS TO REPRODUCE
Add a rule which clones a git repository containing a file called 'origin' in the toplevel directory.

Example rule:
```yaml
- name: Install / update technic mod
  git:
    repo: https://github.com/minetest-technic/technic
    dest: /var/games/minetest-server/.minetest/mods/technic
    update: yes
```

##### EXPECTED RESULTS
The git repository should be cloned on the managed node.

##### ACTUAL RESULTS
ansible aborts with this error message:
```
TASK [program_minetest : Install / update technic mod] *************************
fatal: [nodename]: FAILED! => {"changed": false, "cmd": "/usr/bin/git reset --hard origin", "failed": true, "msg": "Failed to checkout branch master", "rc": 128, "stderr": "fatal: ambiguous argument 'origin': both revision and filename\nUse '--' to separate paths from revisions, like this:\n'git <command> [<revision>...] -- [<file>...]'\n", "stdout": "", "stdout_lines": []}
	to retry, use: --limit @/home/user/Projects/ansible_site/site.retry
```

**Comment by Miyurz** [comment]
@silwol I tested this on the dev branch (version = 2.3.0 0.0.devel)  and this looks more of a git issue.
You can reproduce this with git  by :

1. Create a new file called origin in /var/games/minetest-server/.minetest/mods/technic
2. do a `git reset --hard origin`

We could pass remote as origin/master in git.py module in the source code to fix this and assume the end users would never have such a file name but not sure if thats the real fix. For now, I can suggest you to pass remote on your own and name it other than origin. may be call it as repo/origin or better origin/master if you don't intend to switch the branch and get going. I tested it with the below code and it works just fine!

---

```
  tasks:
    - name: https://github.com/ansible/ansible/issues/22455
      git:
        repo: https://github.com/minetest-technic/technic
        dest: /var/games/minetest-server/.minetest/mods/technic
        remote: origin/master
        update: yes
      register: status
```

**Comment by Miyurz** [comment]
Adding screenshot to prove its git which actually spews the erroneous message.
<img width="590" alt="screen shot 2017-03-10 at 3 40 17 pm" src="https://cloud.githubusercontent.com/assets/2662752/23791205/24ed5110-05a8-11e7-81fe-a84f3033bb3a.png">

<img width="448" alt="screen shot 2017-03-10 at 3 30 07 pm" src="https://cloud.githubusercontent.com/assets/2662752/23791264/53ac6626-05a8-11e7-9923-f89b0d54a172.png">

**Comment by silwol** [comment]
The problem is, that git doesn't know wheter to read the origin as the revision or the filename. Simply calling `git reset --hard origin --` instead of `git reset --hard origin` would resolve the issue, because then it's clear for git that `origin` means the revision.

**Comment by Miyurz** [comment]
@silwol Tested it with https://github.com/ansible/ansible/pull/22502 and it works fine! Lets see what the maintainers have to say.

**Comment by robinro** [comment]
resolved_by_pr #22502

**Comment by Miyurz** [comment]
@gundalow Can you please close this ?

**Comment by sivel** [comment]
Closing as requested.  Resolved by #22502
